Libplanet/KeyStore/ProtectedPrivateKey.cs

```diff
@@ -0,0 +1,402 @@
+        /// <paramref name="passphrase"/>.
+        /// </summary>
+        /// <param name="passphrase">A user input passphrase (i.e., password).</param>
+        /// <returns>A bare <see cref="PrivateKey"/>.</returns>
+        [Pure]
+        public PrivateKey Unprotect(string passphrase)
+        {
+            ImmutableArray<byte> derivedKey = Kdf.Derive(passphrase);
+            var mac = CalculateMac(derivedKey, Ciphertext);
+            if (!Mac.SequenceEqual(mac))
+            {
+                throw new IncorrectPassphraseException(
+                    "The input passphrase is incorrect.",
+                    nameof(passphrase),
+                    Mac,
+                    mac
+                );
+            }
+
+            ImmutableArray<byte> encKey = MakeEncryptionKey(derivedKey);
+            ImmutableArray<byte> plaintext = Cipher.Decrypt(encKey, Ciphertext);
+
+            var key = new PrivateKey(plaintext.ToBuilder().ToArray());
+            Address actualAddress = key.PublicKey.ToAddress();
+            if (!Address.Equals(actualAddress))
+            {
+                throw new MismatchedAddressException(
+                    "The actual address of the unprotected private key does not match to " +
+                    "the expected address of the protected private key.",
+                    Address,
+                    actualAddress
+                );
+            }
+
+            return key;
+        }
+
+        /// <summary>
+        /// Dumps the cipher parameters as a JSON representation according to Ethereum's
+        /// <a href="https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition">Web3
+        /// Secret Storage Definition</a>.
+        /// </summary>
+        /// <param name="writer">A JSON writer which has not begun object nor array.</param>
+        /// <param name="id">A unique identifier, which goes to the <c>id</c> field in the key JSON
+        /// file.  If <c>null</c> (which is default) it is random-generated.</param>
+        public void WriteJson(Utf8JsonWriter writer, [Pure] in Guid? id = null)
+        {
+            writer.WriteStartObject();
+            writer.WriteNumber("version", 3);
+            writer.WriteString("id", (id ?? Guid.NewGuid()).ToString().ToLower());
+            writer.WriteString("address", Address.ToHex().ToLower());
+            writer.WriteStartObject("crypto");
+            writer.WriteString("ciphertext", ByteUtil.Hex(Ciphertext));
+            writer.WritePropertyName("cipherparams");
+            string cipherName = Cipher.WriteJson(writer);
+            writer.WriteString("cipher", cipherName);
+            writer.WritePropertyName("kdfparams");
+            string kdfName = Kdf.WriteJson(writer);
+            writer.WriteString("kdf", kdfName);
+            writer.WriteString("mac", ByteUtil.Hex(Mac));
+            writer.WriteEndObject();
+            writer.WriteEndObject();
```

Comment:
Is there a reason to call `WriteEndObject` twice?

Reply:
Because we have two`WriteStartObject()` calls above?